File: Assets/Typescript~/src/Shared/Entity/Character/Accessory/CharacterAccessories.ts
import { Entity } from "Shared/Entity/Entity";
import { Network } from "Shared/Network";
import { Bin } from "Shared/Util/Bin";
import { RunUtil } from "Shared/Util/RunUtil";
import { Signal } from "Shared/Util/Signal";
import { AccessorySlot } from "./AccessorySlot";
import { AccessoryType } from "./AccessoryType";

export class CharacterAccessories {
	public OnAccessoryRemoved = new Signal<AccessoryType>();
	public OnAccessoryAdded = new Signal<AccessoryType>();

	private accessories = new Map<AccessorySlot, string>();
	private bin = new Bin();

	private finishedFirstPass = false;

	constructor(private entity: Entity) {
		if (RunUtil.IsClient()) {
			this.bin.Add(
				Network.ServerToClient.SetAccessory.Client.OnServerEvent((entityId, slot, path) => {
					if (entityId === this.entity.id) {
						this.SetAccessory(slot, path);
					}
				}),
			);
			this.bin.Add(
				Network.ServerToClient.RemoveAccessory.Client.OnServerEvent((entityId, slot) => {
					if (entityId === this.entity.id) {
						this.RemoveAccessory(slot);
					}
				}),
			);
		}

		this.finishedFirstPass = true;
	}

	public SetAccessory(slot: AccessorySlot, accessoryPath: string): void {
		// const meta = GetAccessoryMeta(accessoryType);
		// let existing = this.accessories.get(slot);
		// if (existing) {
		// 	this.OnAccessoryRemoved.Fire(existing);
		// 	if (RunUtil.IsServer() && this.finishedFirstPass) {
		// 		Network.ServerToClient.RemoveAccessory.Server.FireAllClients(this.entity.Id, accessoryType);
		// 	}
		// }
		// this.accessories.set(meta.category, accessoryType);
		// this.OnAccessoryAdded.Fire(accessoryType);
		// if (RunUtil.IsServer() && this.finishedFirstPass) {
		// 	Network.ServerToClient.AddAccessory.Server.FireAllClients(this.entity.Id, accessoryType);
		// }
	}

	public RemoveAccessory(slot: AccessorySlot): void {
		// const meta = GetAccessoryMeta(accessoryType);
		// let existing = this.accessories.get(meta.category);
		// if (existing === accessoryType) {
		// 	this.OnAccessoryRemoved.Fire(accessoryType);
		// 	this.accessories.delete(meta.category);
		// }
	}

	public Unload(): void {
		this.bin.Clean();
	}

	public GetAccessories(): AccessoryType[] {
		return [];
		// return ObjectUtils.values(this.accessories);
	}
}

File: Assets/Bundles/Shared/Resources/TS/Entity/Character/Accessory/CharacterAccessories.lua
-- Compiled with unity-ts v2.1.0-75
local Network = require("Shared/TS/Network").Network
local Bin = require("Shared/TS/Util/Bin").Bin
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local Signal = require("Shared/TS/Util/Signal").Signal
local CharacterAccessories
do
	CharacterAccessories = setmetatable({}, {
		__tostring = function()
			return "CharacterAccessories"
		end,
	})
	CharacterAccessories.__index = CharacterAccessories
	function CharacterAccessories.new(...)
		local self = setmetatable({}, CharacterAccessories)
		return self:constructor(...) or self
	end
	function CharacterAccessories:constructor(entity)
		self.entity = entity
		self.OnAccessoryRemoved = Signal.new()
		self.OnAccessoryAdded = Signal.new()
		self.accessories = {}
		self.bin = Bin.new()
		self.finishedFirstPass = false
		if RunUtil:IsClient() then
			self.bin:Add(Network.ServerToClient.SetAccessory.Client:OnServerEvent(function(entityId, slot, path)
				if entityId == self.entity.id then
					self:SetAccessory(slot, path)
				end
			end))
			self.bin:Add(Network.ServerToClient.RemoveAccessory.Client:OnServerEvent(function(entityId, slot)
				if entityId == self.entity.id then
					self:RemoveAccessory(slot)
				end
			end))
		end
		self.finishedFirstPass = true
	end
	function CharacterAccessories:SetAccessory(slot, accessoryPath)
	end
	function CharacterAccessories:RemoveAccessory(slot)
	end
	function CharacterAccessories:Unload()
		self.bin:Clean()
	end
	function CharacterAccessories:GetAccessories()
		return {}
		-- return ObjectUtils.values(this.accessories);
	end
end
return {
	CharacterAccessories = CharacterAccessories,
}
-- ----------------------------------
-- ----------------------------------
